### PR TITLE
Implement the delete address endpoint

### PR DESCRIPTION
**Pull request by gaborh-da**

### CODE REVIEW

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java [hunk 1]
```diff
@@ -69,5 +72,69 @@ void GIVEN_balances_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_corr
...
+
+    @Test
+    void GIVEN_only_liquid_balance_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_correct_balances() throws InvalidProtocolBufferException {
+        // GIVEN 1 balance, 2 locked balances, 2 incoming balances
+        double liquidAmount = 100.0;
+        var c1 = BalanceTestUtil.populateBalance(liquidAmount, BalanceTestUtil.IBAN1, BalanceTestUtil.ASSET_CODE1, SANDBOX, getCurrentBankPartyId(), Balance.TEMPLATE_ID);
+
+        // WHEN
+        List<com.rln.gui.backend.model.Balance> balances = RestAssured.given()
+            .when().get(String.format("/api/addresses/%s/balance", BalanceTestUtil.IBAN1))
```
Comment (tamaskalcza-da) [inline gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java:91]:
```suggestion
        List<com.rln.gui.backend.model.Balance> balances = RestAssured
            .get(String.format("/api/addresses/%s/balance", BalanceTestUtil.IBAN1))
```

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java [hunk 2]
```diff
@@ -69,5 +72,69 @@ void GIVEN_balances_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_corr
...
+        // THEN
+        MatcherAssert.assertThat(balances.get(0).getBalance().doubleValue(), Matchers.is(liquidAmount));
+
+        TransactionsApiImplTest.cleanup(getCurrentBankPartyId(), Balance.TEMPLATE_ID, c1.getValue());
+    }
+
+    @Test
+    void GIVEN_balance_not_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_correct_balances() {
+        RestAssured.given()
+            .when().get(String.format("/api/addresses/%s/balance", BalanceTestUtil.IBAN2))
```
Comment (tamaskalcza-da) [inline gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java:106]:
```suggestion
        RestAssured
            .get(String.format("/api/addresses/%s/balance", BalanceTestUtil.IBAN2))
```

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java [hunk 3]
```diff
@@ -69,5 +72,69 @@ void GIVEN_balances_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_corr
...
+        // GIVEN 1 balance, 2 locked balances, 2 incoming balances
+        double liquidAmount = 0.0;
+        BalanceTestUtil.populateBalance(liquidAmount, BalanceTestUtil.IBAN1, BalanceTestUtil.ASSET_CODE1, SANDBOX, getCurrentBankPartyId(), Balance.TEMPLATE_ID);
+
+        // WHEN
+        RestAssured.given()
+            .when().delete(String.format("/api/ledger/addresses/%s", BalanceTestUtil.IBAN1))
+            .then()
+            .assertThat()
+            .statusCode(204);
```
Comment (tamaskalcza-da) [inline gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java:123]:
```suggestion
        RestAssured
            .delete(String.format("/api/ledger/addresses/%s", BalanceTestUtil.IBAN1))
            .then()
            .statusCode(204);
```

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java [hunk 4]
```diff
@@ -69,5 +72,69 @@ void GIVEN_balances_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_corr
...
+        // GIVEN 1 balance, 2 locked balances, 2 incoming balances
+        double liquidAmount = 100.0;
+        var c1 = BalanceTestUtil.populateBalance(liquidAmount, BalanceTestUtil.IBAN1, BalanceTestUtil.ASSET_CODE1, SANDBOX, getCurrentBankPartyId(), Balance.TEMPLATE_ID);
+
+        // WHEN
+        RestAssured.given()
+            .when().delete(String.format("/api/ledger/addresses/%s", BalanceTestUtil.IBAN1))
+            .then()
+            .assertThat()
+            .statusCode(403);
```
Comment (tamaskalcza-da) [inline gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java:137]:
```suggestion
        RestAssured
            .delete(String.format("/api/ledger/addresses/%s", BalanceTestUtil.IBAN1))
            .then()
            .statusCode(403);
```

File: common/src/main/java/com/rln/client/damlClient/RLNDamlClient.java [hunk 5]
```diff
@@ -176,7 +179,14 @@ public void subscribeForContinuousEvent(String subscriberParty, Set<Identifier>
             .forEach(consumer);
     }
 
-    // helper functions
+    @Override
+    public void archiveBalance(ArchiveBalanceParameters parameters) {
+        var event = String.format("Archive balance %s (provider: %s)", parameters.getIban(), parameters.getProvider());
+        var update = Balance.byKey(new BalanceKey(parameters.getProvider(), parameters.getIban())).exerciseArchive(new Archive());
+        commandPublisher.onNext(new ClientCommand(event, update, parameters.getProvider(), parameters));
```
[the remark this answers is not in the document]
Comment (gaborh-da) [reply common/src/main/java/com/rln/client/damlClient/RLNDamlClient.java:186]: Then, this should be done to all the different Parameters classes (5-6).

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java [hunk 6]
```diff
@@ -179,7 +179,7 @@ void apiTransactionsGetRejected() throws InvalidProtocolBufferException {
     cleanup(getCurrentBankPartyId(), RejectedTransferProposal.TEMPLATE_ID, rejectedCid);
   }
 
-  private void cleanup(Party partyId, Identifier identifier, String contractId)
+  public static void cleanup(Party partyId, Identifier identifier, String contractId)
```
Comment (tamaskalcza-da) [inline gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java:182]: Why was this necessary?

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/balanceManagement/BalanceCacheTest.java [hunk 7]
```diff
@@ -39,9 +39,8 @@ void GIVEN_several_balance_created_events_WHEN_update_cache_THEN_balance_aggrega
         cache.updateBalanceUponCreation(balanceAccount2, cid3);
 
         // THEN
-        Map<String, BigDecimal> balances = cache.getBalance();
-        MatcherAssert.assertThat(balances.get(BalanceTestUtil.IBAN1).doubleValue(), Matchers.is(amount1InAccount1 + amount2InAccount1));
-        MatcherAssert.assertThat(balances.get(BalanceTestUtil.IBAN2).doubleValue(), Matchers.is(amountInAccount2));
+        MatcherAssert.assertThat(cache.getBalance(BalanceTestUtil.IBAN1).doubleValue(), Matchers.is(amount1InAccount1 + amount2InAccount1));
```
Comment (tamaskalcza-da) [inline gui-backend/src/test/java/com/rln/gui/backend/implementation/balanceManagement/BalanceCacheTest.java:42]: Looks like the indentation is off in this file.

File: gui-backend/src/main/java/com/rln/gui/backend/implementation/balanceManagement/cache/BalanceCache.java [hunk 8]
```diff
@@ -52,7 +52,7 @@ public void updateBalanceUponArchived(ContractId cid) {
     }
 
     // map from iban to total balanceAmount
-    public Map<String, BigDecimal> getBalance() {
+    public BigDecimal getBalance(String address) {
```
[the remark this answers is not in the document]
Comment (gaborh-da) [reply gui-backend/src/main/java/com/rln/gui/backend/implementation/balanceManagement/cache/BalanceCache.java:55]: I have but back then you said `null` should be enough. :)

File: common/src/main/java/com/rln/client/damlClient/RLNDamlClient.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tamaskalcza-da) [reply common/src/main/java/com/rln/client/damlClient/RLNDamlClient.java:186]: If you referred to the proposed design change then yes. However, I would not consider that in this PR, and would discuss the proposed change offline first.

File: gui-backend/src/main/java/com/rln/gui/backend/implementation/balanceManagement/cache/BalanceCache.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tamaskalcza-da) [reply gui-backend/src/main/java/com/rln/gui/backend/implementation/balanceManagement/cache/BalanceCache.java:55]: Fine by me.

File: gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java [hunk 9]
```diff
@@ -41,29 +50,50 @@ public BalancesApiImpl(
...
             .assetName(accountCache.getAssetCode(address));
 
-        var actualBalance = builder
-            .balance(liquidBalanceAmount.add(lockedBalanceAmount))
-            .type(BalanceType.ACTUAL.name()).build();
+        var result = new ArrayList<Balance>(3);
+
+        if (liquidBalanceAmount == null) {
+            throw new IbanNotFoundException(address);
+        }
```
[the remark this answers is not in the document]
Comment (gaborh-da) [reply gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java:67]: With the Optional API, it is a bit different, though.

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java [hunk 10]
```diff
@@ -69,5 +72,69 @@ void GIVEN_balances_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_corr
...
+        TransactionsApiImplTest.cleanup(getCurrentBankPartyId(), LockedBalance.TEMPLATE_ID, c3.getValue());
+        TransactionsApiImplTest.cleanup(getCurrentBankPartyId(), IncomingBalance.TEMPLATE_ID, c4.getValue());
+        TransactionsApiImplTest.cleanup(getCurrentBankPartyId(), IncomingBalance.TEMPLATE_ID, c5.getValue());
+    }
+
+    @Test
+    void GIVEN_only_liquid_balance_on_ledger_WHEN_get_request_balance_endpoint_THEN_return_correct_balances() throws InvalidProtocolBufferException {
+        // GIVEN 1 balance, 2 locked balances, 2 incoming balances
+        double liquidAmount = 100.0;
+        var c1 = BalanceTestUtil.populateBalance(liquidAmount, BalanceTestUtil.IBAN1, BalanceTestUtil.ASSET_CODE1, SANDBOX, getCurrentBankPartyId(), Balance.TEMPLATE_ID);
```
[the remark this answers is not in the document]
Comment (gaborh-da) [reply gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java:87]: renamed to balanceId1, ...

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java (hunk 2 above)
Comment (gaborh-da) [reply gui-backend/src/test/java/com/rln/gui/backend/implementation/BalancesApiImplTest.java:106]: I deleted those

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java (hunk 6 above)
Comment (gaborh-da) [reply gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java:182]: I wanted to use the same method in BalancesApiImplTest.

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/balanceManagement/BalanceCacheTest.java (hunk 7 above)
Comment (gaborh-da) [reply gui-backend/src/test/java/com/rln/gui/backend/implementation/balanceManagement/BalanceCacheTest.java:42]: formatted

File: gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java [hunk 11]
```diff
@@ -41,29 +50,50 @@ public BalancesApiImpl(
...
-        var lockedBalanceAmount = lockedBalanceCache.getBalance().get(address);
+    public List<Balance> getAddressBalance(String address) throws IbanNotFoundException {
+        var liquidBalanceAmount = liquidBalanceCache.getBalance(address);
+        var incomingBalanceAmount = incomingBalanceCache.getBalance(address);
+        var lockedBalanceAmount = lockedBalanceCache.getBalance(address);
 
         var builder = Balance.builder()
             .address(address)
             .assetId(0L) // default to 0 now, as we don't have assetId in the system yet
             .assetName(accountCache.getAssetCode(address));
```
[the remark this answers is not in the document]
Comment (gaborh-da) [reply gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java:61]: > First I found inconsistency: while balance caches return even when address cannot be found, the account cache fails.

Which one would you do to make it consistent?

[the remark this answers is not in the document]
Comment (gaborh-da) [reply gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java:61]: > Second, I would consider BalancesApiImpl web-related, and if that was the case, we could simplify the code by directly throwing web related exceptions here.

Cannot we consider it as the implementation for balance related services but not yet web related / throwing web exceptions? So they would only be thrown in GuiBackendApiImplementation.

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java (hunk 6 above)
Comment (tamaskalcza-da) [reply gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java:182]: Then can we name it properly and move it to `LedgerBaseTest` or some utility class? E.g. `cleanupContract` or `archive`.

File: gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java [hunk 12]
```diff
@@ -41,29 +50,46 @@ public BalancesApiImpl(
      *      ACTUAL: liquid + locked
      *      FUTURE: liquid + incoming
      */
-    public List<Balance> getAddressBalance(String address) {
-        var liquidBalanceAmount = liquidBalanceCache.getBalance().get(address);
-        var incomingBalanceAmount = incomingBalanceCache.getBalance().get(address);
-        var lockedBalanceAmount = lockedBalanceCache.getBalance().get(address);
+    public List<Balance> getAddressBalance(String address) throws IbanNotFoundException {
+        var liquidBalanceAmount = liquidBalanceCache.getBalance(address);
+        var actualLiquidAmount = liquidBalanceAmount.orElseThrow(() -> new IbanNotFoundException(address));
```
Comment (tamaskalcza-da) [inline gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java:55]: How about?
```suggestion
        var liquidBalanceAmount = liquidBalanceCache
          .getBalance(address)
          .orElseThrow(() -> new IbanNotFoundException(address));
```

File: gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (tamaskalcza-da) [reply gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java]: > > Second, I would consider BalancesApiImpl web-related, and if that was the case, we could simplify the code by directly throwing web related exceptions here.
> 
> Cannot we consider it as the implementation for balance related services but not yet web related / throwing web exceptions? So they would only be thrown in GuiBackendApiImplementation.

That's valid as well. We can keep it like this.

[the remark this answers is not in the document]
Comment (tamaskalcza-da) [reply gui-backend/src/main/java/com/rln/gui/backend/implementation/methods/BalancesApiImpl.java]: > > First I found inconsistency: while balance caches return even when address cannot be found, the account cache fails.
> 
> Which one would you do to make it consistent?

The way balance caches behave. I think from their point of view `null` or `Optional<T>` is a perfectly valid response. What these values mean from a business perspective can actually be decided in the client code.

File: gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java (hunk 6 above)
Comment (gaborh-da) [reply gui-backend/src/test/java/com/rln/gui/backend/implementation/TransactionsApiImplTest.java:182]: okay, LedgerBaseTest

**Review by tamaskalcza-da — APPROVED** [review]

**Review by darko-da — APPROVED** [review]
Rubber Stamp LGTM
